src/supervisor/mod.rs

```diff
@@ -198,135 +211,320 @@ async fn terminate_children<'a>(
+            ev_notifier
+                .supervisor_start_failed(sup_runtime_name, err.clone())
+                .await;
+            let spec = Spec {
+                meta,
+                children: spec_children,
+            };
+            let start_result = (spec, err);
+            match mstart_notifier {
+                Some(start_notifier) => {
+                    // if we call this function inside a `task::spawn`, the
+                    // spawner should receive back the spec and error
+                    start_notifier.failed(start_result);
+                    // this implementation was forced by the compiler, start_result cannot
+                    // belong to both a spawner and a caller, rust compile, thank you <3
+                    return Err(None);
+                }
+                None => {
+                    // if we call this function on the same thread, the function
+                    // caller should receive back the spec and error
+                    return Err(Some(start_result));
+                }
+            }
+        }
+        // notify supervisor of start success
+        Ok(sup_runtime_children) => {
+            // here we signal that the start was successful, which will unblock
+            // the bootstrap logic to spawn the next sibling
+            ev_notifier.supervisor_started(&sup_runtime_name).await;
+            if let Some(start_notifier) = mstart_notifier {
+                start_notifier.success(());
+            }
+
+            // after startup, we initialize the monitor loop where we listen
+            // to children errors or public API calls
+            loop {
+                tokio::select! {
+                    // if our context is done, it means the termination_handle
+                    // got called; signaling a shutdown of the whole supervision
+                    // tree.
+                    _ = ctx.done.clone() => {
+                        let termination_result =
+                            terminate_supervisor_monitor(
+                                meta,
+                                sup_runtime_name,
+                                sup_runtime_children,
+                            ).await;
+
+                        // Ok means we did not fail to start, but
+                        // termination_result contains a Result that indicates
+                        // if we failed to terminate or not
+                        return Ok(termination_result);
+                    },
+                    // TODO: monotiring children
+                    // ch_notification = notify_recv.receive() => {
+                    //
+                    // }
+                    // TODO: control cmd listening
+                    // ctrl_msg = ctrl_recv.receive() => {
+                    //
+                    // }
+                }
+            }
+        }
     }
 }
 
 impl Spec {
     pub fn new(
-        name: impl Into<String>,
+        name0: impl Into<String>,
```

Comment:
This can stay `name`. Not Haskell, remember? :)

Reply:
:grimacing: